File: iosApp/iosApp/Interop/DatadogAppLogger.swift
import ComposeApp
import DatadogLogs
import Foundation

/// Bridges the native Datadog iOS `LoggerProtocol` into Kotlin's `UtilsAppLoggerApi`, so shared
/// Kotlin code can log through the app-wide `AppLogger` without Kotlin/Native ever needing to link
/// the Datadog SDK's iOS framework itself (see `installAppLogger` in `IOSLoggerBootstrap.kt`).
final class DatadogAppLogger: NSObject, UtilsAppLoggerApi {
    private let logger: LoggerProtocol

    init(logger: LoggerProtocol) {
        self.logger = logger
    }

    func debug(message: String, throwable: KotlinThrowable?) {
        logger.debug(message, error: throwable?.asError(), attributes: nil)
    }

    func info(message: String, throwable: KotlinThrowable?) {
        logger.info(message, error: throwable?.asError(), attributes: nil)
    }

    func warn(message: String, throwable: KotlinThrowable?) {
        logger.warn(message, error: throwable?.asError(), attributes: nil)
    }

    func error(message: String, throwable: KotlinThrowable?) {
        logger.error(message, error: throwable?.asError(), attributes: nil)
    }
}

File: shared/src/androidMain/kotlin/nl/q42/template/logging/DatadogAppLogger.kt
package nl.q42.template.logging

import com.datadog.kmp.log.Logger
import nl.q42.template.core.utils.logging.AppLoggerApi

internal class DatadogAppLogger(private val logger: Logger) : AppLoggerApi {
    override fun debug(message: String, throwable: Throwable?) = logger.debug(message, throwable)
    override fun info(message: String, throwable: Throwable?) = logger.info(message, throwable)
    override fun warn(message: String, throwable: Throwable?) = logger.warn(message, throwable)
    override fun error(message: String, throwable: Throwable?) = logger.error(message, throwable)
}
